### PR TITLE
Aarondonahue/scream atm mct

### PR DESCRIPTION
**Pull request by AaronDonahue**
Improves the CIME build version of scream such that it now builds and runs scream as the ATM component of E3SM.

1) Treats the ATM component as a DEAD component with respects to how it is viewed by the component coupler.  This commit includes all of the necessary modules for linking to MCT.
2) Within the "DEAD" component for ATM, scream builds the atmosphere-driver (AD).
3) Currently only P3+SHOC are run with an idealized initial condition.  Matches exactly the stand-alone scream-ad test already in the unit test suite.
4) Minor infrastructure and P3 modifications to consistently copy the p3 data tables across tests and within the cime build.  This required some changes to the hard-coded location of the p3 data tables in various locations of the code.

[BFB]

### CODE REVIEW

File: components/scream/cime_config/buildnml [hunk 1]
```diff
@@ -12,7 +12,7 @@ sys.path.append(os.path.join(_CIMEROOT, "scripts", "Tools"))
 from standard_script_setup import *
 from CIME.case import Case
 from CIME.utils import expect, run_cmd_no_fail, safe_copy
-from CIME.buildnml import create_namelist_infile, parse_input
+from CIME.buildnml import create_namelist_infile, parse_input, build_xcpl_nml
```
Comment (jgfouca) [inline components/scream/cime_config/buildnml]: Where is this new import used?

File: components/scream/src/interface/CMakeLists.txt [hunk 2]
```diff
@@ -10,8 +13,10 @@ else()
 endif()
 
 set(CSM_SHARE "${INSTALL_SHAREDPATH}/${COMP_INTERFACE}/${ESMFDIR}/${NINST_VALUE}/csm_share")
+link_directories(${SCREAM_TPL_LIBRARY_DIRS})
 
 add_library(atm ${ATM_SRC})
-target_include_directories(atm PUBLIC ${SCREAM_INCLUDE_DIRS} ${SCREAM_TPL_INCLUDE_DIRS} ${CSM_SHARE} ${INSTALL_SHAREDPATH}/include)
+target_include_directories(atm PUBLIC ${SCREAM_INCLUDE_DIRS} ${SCREAM_TPL_INCLUDE_DIRS} ${CSM_SHARE} ${INSTALL_SHAREDPATH}/include ${CATCH_INCLUDE_DIR})
```
Comment (jgfouca) [inline components/scream/src/interface/CMakeLists.txt]: Why is catch needed for a library?

File: components/scream/src/interface/atm_comp_mct.F90.old [hunk 3]
```diff
@@ -0,0 +1,119 @@
+module atm_comp_mct
```
Comment (bartgol) [inline components/scream/src/interface/atm_comp_mct.F90.old]: Is this file needed? Looks like it was a backup before you started doing mods. If so, let's remove it.

File: components/scream/src/interface/atm_comp_mct.F90.old [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+  ! These are used across the module, so we issue a single use clause here
```
Comment (jgfouca) [inline components/scream/src/interface/atm_comp_mct.F90.old]: A good version-control system like Git should mean we don't need to keep around .old files. Is there something special that means we have to keep this?

File: components/scream/src/physics/p3/CMakeLists.txt [hunk 5]
```diff
@@ -38,6 +38,6 @@ add_library(p3 ${P3_SRCS})
 target_include_directories(p3 PUBLIC ${SCREAM_INCLUDE_DIRS} ${SCREAM_TPL_INCLUDE_DIRS} ${CIMEROOT}/src/share/include)
 set_target_properties(p3 PROPERTIES
   Fortran_MODULE_DIRECTORY ${SCREAM_F90_MODULES})
-# target_link_libraries(p3 scream_share ${SCREAM_TPL_LIBRARIES})
+target_link_libraries(p3 scream_share ${SCREAM_TPL_LIBRARIES})
```
Comment (jgfouca) [inline components/scream/src/physics/p3/CMakeLists.txt]: Just curious what's going on here?

File: components/scream/src/physics/p3/scream_p3_interface.F90 [hunk 6]
```diff
@@ -29,7 +29,8 @@ module scream_p3_interface_mod
   integer(kind=c_int) :: qsize = 9
 
   character(len=16)   :: micro_p3_tableversion = "4"
-  character(len=100)  :: micro_p3_lookup_dir = "./"
+  character(len=100)  :: micro_p3_lookup_dir = "./data"
+  !character(len=100)  :: micro_p3_lookup_dir = "/usr/gdata/climdat/ccsm3data/inputdata/atm/cam/physprops"
```
Comment (jgfouca) [inline components/scream/src/physics/p3/scream_p3_interface.F90]: Remove commented-out line.

File: components/scream/src/interface/scream_cxx_f90_interface.cpp [hunk 7]
```diff
@@ -1,57 +1,125 @@
-#include "interface/ScreamContext.hpp"
+#include <catch2/catch.hpp>
```
Comment (jgfouca) [inline components/scream/src/interface/scream_cxx_f90_interface.cpp]: I don't think this catch include is needed.

File: components/scream/src/interface/atm_comp_mct.F90 [hunk 8]
```diff
@@ -36,36 +63,104 @@ subroutine scream_init(f_comm,start_ymd,start_tod) bind(c)
...
+
+    call shr_file_getLogUnit (shrlogunit)
+    call shr_file_getLogLevel(shrloglev)
+    call shr_file_setLogUnit (logUnit)
+
+    !----------------------------------------------------------------------------
+    ! Initialize xatm
+    !----------------------------------------------------------------------------
+
+    call dead_init_mct('atm', Eclock, x2d, d2x, &
```
Comment (bartgol) [inline components/scream/src/interface/atm_comp_mct.F90]: Are the dead_xyz calls there only for now, to get things working, or do they perform some common operations (common meaning needed in both a dead_atm setup and an active_atm setup)?

If they will always be used, we should rename the dead_xyz to remove 'dead'. If they are used only in dead_atm runs, we should guard them with some compile-time ifdef stuff. Or, at the very least, we should put some comment that remind the user that this stuff is there only for dead_atm runs, and it should be skipped/avoided in active_atm runs.

File: components/scream/src/physics/p3/CMakeLists.txt (hunk 5 above)
Comment (AaronDonahue) [reply components/scream/src/physics/p3/CMakeLists.txt]: To be honest I am not a CMake expert, so I'm not sure.  But when I remove (or comment) this line the build in CIME breaks.

File: components/scream/src/physics/p3/scream_p3_interface.F90 [hunk 9]
```diff
@@ -70,14 +71,15 @@ subroutine p3_standalone_init_f90 (q,T,zi,pmid,pdel,ast,naai,npccn) bind(c)
...
-    character(len=100) :: case_title
+    character(len=100) :: case_title, tmp_c1, tmp_c2
 
     integer(kind=c_int) :: i, k
     logical(kind=c_bool) :: masterproc
 
     ! READ inputs from SCM for p3-stand-alone:
     q(:,:,:) = 0.0_rtype
     open(unit=981,file='./data/p3_universal_constants.inp',status='old',action='read')
+    !open(unit=981,file='../build/cmake-bld/scream/tests/scream_p3/data/p3_universal_constants.inp',status='old',action='read')
```
Comment (bartgol) [inline components/scream/src/physics/p3/scream_p3_interface.F90]: If possible, let's not add commented code.

File: components/scream/src/physics/p3/scream_p3_interface.F90 (hunk 6 above)
Comment (AaronDonahue) [reply components/scream/src/physics/p3/scream_p3_interface.F90]: yes, good catch

File: components/scream/src/interface/scream_cxx_f90_interface.cpp (hunk 7 above)
Comment (AaronDonahue) [reply components/scream/src/interface/scream_cxx_f90_interface.cpp]: I'll remove it and see what happens.

File: components/scream/src/interface/atm_comp_mct.F90.old (hunk 4 above)
Comment (AaronDonahue) [reply components/scream/src/interface/atm_comp_mct.F90.old]: yes, I'll delete the file.

File: components/scream/src/interface/CMakeLists.txt (hunk 2 above)
Comment (AaronDonahue) [reply components/scream/src/interface/CMakeLists.txt]: Not sure, I think I copied this line from somewhere else.  I will remove the catch.

File: components/scream/cime_config/buildnml (hunk 1 above)
Comment (AaronDonahue) [reply components/scream/cime_config/buildnml]: I'm not sure.  It looks like xcpl_nml might refer to the "dead" component stuff but I can't remember putting this in.  I may have been copying from an example.  I'll remove this and see if it still works.

File: components/scream/src/physics/p3/CMakeLists.txt (hunk 5 above)
Comment (bartgol) [reply components/scream/src/physics/p3/CMakeLists.txt]: My guess: some executable was not explicitly asking to link to scream_share. Honestly, if p3 uses scream_share (and it does), it is cleaner to have p3 "linking" scream_share. It doesn't really link (it's a library), but its dependencies are then transferred to downstream targets. This is better than having a customer of libp3.a having to figure out all of p3 deps and link them.

**tl;dr**: I think it's correct to call `target_link_libraries`.

File: components/scream/src/interface/CMakeLists.txt (hunk 2 above)
Comment (bartgol) [reply components/scream/src/interface/CMakeLists.txt]: I think it's because you were including catch in `scream_cxx_f90_interface.cpp `. Now that you removed it, you can revert this change.

File: components/scream/src/physics/p3/CMakeLists.txt (hunk 5 above)
Comment (AaronDonahue) [reply components/scream/src/physics/p3/CMakeLists.txt]: So I should leave this in?

Comment (bartgol) [reply components/scream/src/physics/p3/CMakeLists.txt]: The way it is right now is correct, I think.

File: components/scream/src/interface/atm_comp_mct.F90 (hunk 8 above)
Comment (AaronDonahue) [reply components/scream/src/interface/atm_comp_mct.F90]: Yes, they are essentially place holders for now.  Once we are ready to turn ATM on I think we will be changing them.  Since I'm not yet an expert on the MCT coupling I'm inclined to leave them as is.

File: components/scream/src/physics/p3/scream_p3_interface.F90 (hunk 9 above)
Comment (AaronDonahue) [reply components/scream/src/physics/p3/scream_p3_interface.F90]: For sure, I'll delete.

**Comment by AaronDonahue** [comment]
Looks like all green, I'm going to merge.
